### PR TITLE
groups: allow users to block other users from DMing them

### PR DESCRIPTION
**Pull request by patosullivan**
Fixes LAND-914, adds a `blocked` set of ships to state on %chat and associated marks/pokes/scry/UI elements.

If we hear a dm-action update from a blocked ship we ignore it.

https://github.com/tloncorp/landscape-apps/assets/1221094/7bd7a0e2-5889-4c4e-acb2-e81428d40374

### CODE REVIEW

File: desk/app/chat.hoon [hunk 1]
```diff
@@ -358,6 +400,8 @@
     ::  don't proxy to self, creates an infinite loop
     ?:  =(p.action our.bowl)
       ~|("%dm-action poke failed: can't dm self" !!)
+    ?:  (~(has in blocked) src.bowl)
+      ~|("%dm-action poke failed: ship blocked" !!)
```
Comment (arthyn) [inline desk/app/chat.hoon:404]: need to add this check to `di-take-counter` which is where we normally receive dms from other parties. this route is more for ourselves.

File: desk/app/chat.hoon [hunk 2]
```diff
@@ -774,6 +835,8 @@
     [%x %clubs ~]  ``clubs+!>((~(run by clubs) |=(=club:c crew.club)))
   ::
     [%x %pins ~]  ``chat-pins+!>(pins)
+  ::
+    [%x %blocked ~]  ``chat-blocked-ships+!>(blocked)
```
Comment (arthyn) [inline desk/app/chat.hoon:839]:
```suggestion
    [%x %blocked ~]  ``ships+!>(blocked)
```

File: desk/mar/chat/blocked-ships.hoon [hunk 3]
```diff
@@ -0,0 +1,15 @@
+/-  c=chat
```
Comment (arthyn) [inline desk/mar/chat/blocked-ships.hoon:1]: can delete this guy we already have a ships mark

**Review by arthyn — CHANGES_REQUESTED** [review]

File: desk/app/chat.hoon (hunk 1 above)
Comment (arthyn) [reply desk/app/chat.hoon:404]: oh nvm I see you put it later

**Comment by zalberico** [comment]
From Sync, example of unblock profile from twitter:
Screenshot 2023-09-29 at 8 23 30 AM

They also have a list of blocked users to unblock in settings.

It's also important for the person who is blocked to know they can't dm (helps reduce abuse - if someone knows they can't message on the platform them they won't, if they think they can then things can escalate with them getting crazier thinking the person is seeing their messages).

**Comment by patosullivan** [comment]
I added an unblock button to the "You have blocked this user" text shown in place of the chat input: 

image

**Comment by patosullivan** [comment]
https://github.com/tloncorp/landscape-apps/assets/1221094/cccbd4f6-6bb1-4bbb-a4f5-322fdca1b513

**Review by arthyn — APPROVED** [review]
lgtm this is great

**Comment by patosullivan** [comment]
Updated to allow users to know when they've been blocked.


https://github.com/tloncorp/landscape-apps/assets/1221094/f65e7975-b1cd-4f13-8dad-3c59770914fa

**Comment by jamesacklin** [comment]
Fantastic, thank you.

File: ui/src/profiles/ProfileModal.tsx [hunk 4]
```diff
@@ -96,6 +122,15 @@ export default function ProfileModal() {
             <FavoriteGroupGrid groupFlags={contact.groups} />
           </div>
         )}
+        {shipHasBlockedUs && (
+          <div className="mt-5">
+            <h2 className="mb-3 font-semibold">Blocked You</h2>
```
Comment (jamesacklin) [inline ui/src/profiles/ProfileModal.tsx:127]: Let's just use "Blocked"

**Review by jamesacklin — APPROVED** [review]

**Review by arthyn — APPROVED** [review]
lgtm

File: desk/lib/chat-json.hoon [hunk 5]
```diff
@@ -385,6 +397,16 @@
...
+  ++  blocked-by-ship
+    %-  ot
+    :~  blocked-by/(se %p)
+    ==
+  ::
+  ++  unblocked-by-ship
+    %-  ot
+    :~  unblocked-by/(se %p)
+    ==
+  ::
```
[the remark this answers is not in the document]
Comment (patosullivan) [reply desk/lib/chat-json.hoon:409]: I think it's necessary for the subscription facts sent to the frontend, right?

[the remark this answers is not in the document]
Comment (patosullivan) [reply desk/lib/chat-json.hoon:409]: nvm I just realized you were commenting on the de-js functions.

**Pull request merged**